### DISCUSSION TITLE
Does the IDE container perform data persistence?

### DISCUSSION THREAD

**Discussion by wyhdil** [eclipse-theia/theia-cloud#246] [Q&A] [answered]
Hello,I have some questions about data persistence that I would like to ask.

I logged in using an account and opened an IDE pod. I modified the data in readme.md.
![9a97923407826dae874b24446a491f1d](https://github.com/eclipsesource/theia-cloud/assets/69418800/4856ef51-60cc-4d5d-aebf-801a86ca3240)

But when I closed the browser and made sure that the pod was automatically deleted, when I logged in again using this account, I found that the previous modifications did not take effect.
![52a1cf7fec220164072b511046553322](https://github.com/eclipsesource/theia-cloud/assets/69418800/d51662da-5f47-4cfa-b5c6-6ea694d79dac)

Does this mean that data persistence is not implemented?
I hope that this status can be maintained after the user uses the IDE to modify the project code.What should I do？
Hope you can help me!  Thanks!!!

**Comment by sgraband** [comment]
Hi,

data persistence is implemented, however it also needs to be configured. Since i have no knowledge about your deployment i would need to guess, what the issue here is, or what you are missing from a configuration side.
Could you check:
- that your Launch Request for the new session is correct (e.g. look [at these fields](https://github.com/eclipsesource/theia-cloud/blob/685e1eca2f3b78b2f4882a54f03ffd6076f4a1e9/java/service/org.eclipse.theia.cloud.service/src/main/java/org/eclipse/theia/cloud/service/LaunchRequest.java#L27-L34))
- if your [values](https://github.com/eclipsesource/theia-cloud-helm/blob/2c7fe9eb083c4223fafafea1967df3924a4d5088/charts/theia.cloud/values.yaml#L201-L206) are configure correctly (e.g. requested storage > 0 and storageClassName is available on your cluster
- and lastly, you could check, whether or not a new workspace resource is created, when you start a new session with the same user. If this is the case your landing-page might request a new session and does not reuse existing ones.

Hope this helps.

**Please note:** This is a free support forum. While we're here to help and answer questions, our time for in-depth debugging is limited. If you require extensive assistance, please consider [our support package](https://eclipsesource.com/services/developer-support/). Your understanding is appreciated.

Thank you for being a part of our community!

**Reply by wyhdil** [reply]
Hello, thank you for your response. I have already checked these options, and it seems that there shouldn't be any issues. Here is my configuration:

This is the request body when the landing page sends a request. The values for workspaceName and ephemeral should be correct.
Additionally, I haven't made any changes to the Theia Cloud configuration, so the configuration you mentioned in the second step should also be correct.
<img width="1629" alt="image" src="https://github.com/eclipsesource/theia-cloud/assets/69418800/7642cc71-5688-4c06-88e8-339979bb4c1b">

Meanwhile, I have also found the workspace CR.
<img width="1058" alt="image" src="https://github.com/eclipsesource/theia-cloud/assets/69418800/590223f7-8718-4387-8771-1579da9cea2d">

This is the pod for the IDE.
```
  theia-cloud-demo:
    Container ID:   docker://b585a8338b7baa5626ed4eb53cea8fb2b4383a325918896ec000954cab9ca176
    Image:          theiacloud/theia-cloud-demo:0.8.1.OSWeek23-v1
    Image ID:       docker://sha256:5443488d49376b502b5db90ef4a13e998cd1b229c1ac301981a84e1be9e887c3
    Port:           3000/TCP
    Host Port:      0/TCP
    State:          Running
      Started:      Thu, 26 Oct 2023 10:18:17 +0800
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     2
      memory:  1200M
    Requests:
      cpu:     100m
      memory:  1G
    Environment:
      THEIACLOUD_APP_ID:                           asdfghjkl
      THEIACLOUD_SERVICE_URL:                      https://192.168.49.2.nip.io/servicex
      THEIACLOUD_SESSION_UID:                      1bbefce7-aec1-4aaa-9a74-cff56cbdb357
      THEIACLOUD_SESSION_NAME:                     ws-asdfghjkl-theia-cloud-demo-admin-example-com-session
      THEIACLOUD_SESSION_USER:                     admin@example.com
      THEIACLOUD_SESSION_URL:                      https://192.168.49.2.nip.io/instances/1bbefce7-aec1-4aaa-9a74-cff56cbdb357/
      THEIACLOUD_SESSION_SECRET:                   8f3e44b7-d1a2-40b4-aa4d-854de69d97d2
      THEIACLOUD_MONITOR_ENABLE_ACTIVITY_TRACKER:  placeholder-enable-activity-tracker
      THEIACLOUD_MONITOR_PORT:                     placeholder-monitor-env-port
      THEIACLOUD_KEYCLOAK_URL:                     https://192.168.49.2.nip.io/keycloak/
      THEIACLOUD_KEYCLOAK_REALM:                   TheiaCloud
      THEIACLOUD_KEYCLOAK_CLIENT_ID:               theia-cloud
    Mounts:
      /home/project/persisted from user-data (rw)
Conditions:
  Type              Status
  Initialized       True 
  Ready             True 
  ContainersReady   True 
  PodScheduled      True 
Volumes:
  oauth2-proxy-config:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      a1bbefce7-aec1-4aaa-9a74-cff56cbdb357-config-admin-example-com
    Optional:  false
  oauth2-templates:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      oauth2-templates
    Optional:  false
  oauth2-emails:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      a1bbefce7-aec1-4aaa-9a74-cff56cbdb357-emailconfig-admin-exampl
    Optional:  false
  user-data:
    Type:        PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:   ba61a342-f68c-4bfb-9f5b-cb4f5a55e6ee-pvc-ws-asdfghjkl-theia-cl
    ReadOnly:    false
QoS Class:       Burstable
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                 node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type    Reason     Age   From               Message
  ----    ------     ----  ----               -------
  Normal  Scheduled  90s   default-scheduler  Successfully assigned theiacloud/a1bbefce7-aec1-4aaa-9a74-cff56cbdb357-deployment-admin-exahln6m to minikube
  Normal  Pulled     90s   kubelet            Container image "theiacloud/theia-cloud-wondershaper:0.8.1.OSWeek23-v1" already present on machine
  Normal  Created    90s   kubelet            Created container wondershaper-init
  Normal  Started    90s   kubelet            Started container wondershaper-init
  Normal  Pulling    89s   kubelet            Pulling image "quay.io/oauth2-proxy/oauth2-proxy:v7.4.0"
  Normal  Pulled     86s   kubelet            Successfully pulled image "quay.io/oauth2-proxy/oauth2-proxy:v7.4.0" in 3.547376512s (3.547383395s including waiting)
  Normal  Created    86s   kubelet            Created container oauth2-proxy
  Normal  Started    85s   kubelet            Started container oauth2-proxy
  Normal  Pulled     85s   kubelet            Container image "theiacloud/theia-cloud-demo:0.8.1.OSWeek23-v1" already present on machine
  Normal  Created    85s   kubelet            Created container theia-cloud-demo
  Normal  Started    85s   kubelet            Started container theia-cloud-demo
```

I'm not sure if there are any additional configurations that need to be done. Currently, it seems that the desired data storage effect hasn't been achieved. The IDE pod is automatically deleted, and when using the same user login, the previous code modifications are not saved. Each time, it starts with a new pod and the same file directory, which is not the desired outcome. I would like to know what else needs to be done to implement this functionality.

Thank you very much for your patience in answering my questions!

**Reply by sgraband** [reply]
The file you are working on is outside of the mounted folder. You are working on `home/project/README.md`, but the mounted volume (as seen in the pod overview) is `home/project/persisted`. So everything in the `persisted` folder should be persisted. Could you try that out?

**Reply by wyhdil** [reply]
Hello,

I found the /home/project/persisted folder but failed to create new file in it since its owner is root:

![img_v2_52ed8a00-de52-48c9-823a-0af3ca7518dg](https://github.com/eclipsesource/theia-cloud/assets/69418800/048f9863-1151-4b03-9164-7d4a356472f6)

I tried to find the creation of the folder and found the theia-cloud-demo image in the dockerhub:

https://hub.docker.com/layers/theiacloud/theia-cloud-demo/0.8.0.MS11/images/sha256-d04f5433c7e2a4aa8e9a394dfe7a64fca2ea54fcf4582503bfcb1e9eae6be481?context=explore

But it does not contain the creation of the folder /home/project/persisted. Does the image use the [Dockerfile](https://github.com/eclipse-theia/theia-blueprint/blob/master/browser.Dockerfile) of the theia-blueprint repository?

What the point is, the owner of the folder /home/project/persisted should be theia:theia to store users codes.

**Reply by sgraband** [reply]
The folder should not be created during the docker build. It is mounted by Theia Cloud into the docker image after the fact. Take a look[ here](https://github.com/eclipsesource/theia-cloud/blob/3d14e85b674285f7ea2d76cdf9503027a9709d69/java/operator/org.eclipse.theia.cloud.operator/src/main/java/org/eclipse/theia/cloud/operator/handler/impl/LazySessionHandler.java#L446-L461).

Note, that the theia-cloud-demo was designed as a minimal example without persistance, so there is no claim of it working with persistence out of the box.

You could check the following:
- Get the uid of the theia user: `id -u theia` (in the console of a session).
- Check that this is the uid in the AppDefinition that you are using.
- If those two match you can also check the deployment of your session. 

For this create a new sessions and then try `kubectl get deployments -n <namespace`.
Get the deployment that was created for your session: `kubectl get deployment <deploymentName> -n <namespace> -o yaml`.
Then take a look at the file and see if the `spec.containers.env.securityContext` contains your UID for runAsGroup and runAsUser.
Furthermore you can check if the `spec.containers.env.volumeMounts` is actually correct.

**Reply by wyhdil** [reply]
Hello, if the folder should not be created during the docker build. Does that mean that the persistence of this folder has no way to play a role, because there is no way to use this folder to write code? And I checked what you said.
1.uid of the theia user
<img width="1166" alt="image" src="https://github.com/eclipsesource/theia-cloud/assets/69418800/b45aed9e-18de-4174-b791-ba790f2f590f">

2.uid in AppDefinition
![ec69f198111d2d46dba936738b02cd63](https://github.com/eclipsesource/theia-cloud/assets/69418800/a185e667-e77c-4c17-ab78-fcf719553396)

3.spec.containers.env.securityContext and spec.containers.env.volumeMounts
![2b3ddf4fca3a3e5c5e116fb011dacd29](https://github.com/eclipsesource/theia-cloud/assets/69418800/4ec18624-5166-4291-993e-5543c70a1984)

Judging from these, are there any misconfigurations?

**Reply by sgraband** [reply]
No the configuration seems to be correct. The arrow in the second screenshot should point to the UID in the spec, though, but those are all the same user. So this is fine.

> if the folder should not be created during the docker build. Does that mean that the persistence of this folder has no way to play a role, because there is no way to use this folder to write code? 

I am not really sure, what you mean by that question? During the Docker build a _Docker image_ is being created. You can run this _image_ to get a _Docker container_. This is done by Theia Cloud, when you start a session. During the creation of the _container_ Theia Cloud will also mount some storage into the _container_. This data comes from a _Kubernetes Persistent Volume_. In theory if you change the data inside that folder, this should be represented in the  _Persistent Volume_. For some reason that is not happening for you.

If the folder would be created using the docker image, there would be no persistence, as everytime a new container comes up the same instructions are executed. Maybe this is the problem? Does you application docker image create a folder under `home/project/persisted`?

**Reply by wyhdil** [reply]
No, I'm just using the command line within the container to create a new directory, but it seems like I can't perform certain operations within this directory.
![5afc689c7d426e2d47b358b664eeb755](https://github.com/eclipsesource/theia-cloud/assets/69418800/0b1e010b-2dc1-4f5e-a11c-55403bf99632)
If operations cannot be performed within this folder, doesn't it render the purpose of the persistent volume obsolete?

**Reply by sgraband** [reply, marked as answer]
Unfortunately i am not sure why the persisted folder belongs to the root user. You could try to debug at which point the folder is created and who exactly creates it. I would assume, that it is not being mounted into the container by kubernetes, because that seems to be setup correctly. Did you check the logs of the pod to see if the mounting is successful?

**Reply by wyhdil** [reply]
Ok, i'll check it.

**Reply by wyhdil** [reply]
Hello, I've encountered another peculiar issue, and I hope you can help take a look.

Here is the storage size for my custom PV.
![38ddc6e5e9ee3fb72148cb39e4f102ec](https://github.com/eclipsesource/theia-cloud/assets/69418800/fada3380-6f45-47d0-b573-58d4136f2a9c)
Clearly, I have set the storage size to 10M.

However, when I create a file of 5.1GB size in the directory mounted by this PVC, it seems to be successful, even though it clearly exceeds the defined storage size of 10M.
![image](https://github.com/eclipsesource/theia-cloud/assets/69418800/3fc14e97-3fbe-42f9-879c-62f3ea2badc2)

It's obviously very strange; the size of the created file far exceeds the defined storage size.
From this perspective, it seems that setting the storage size for the PVC has no effect whatsoever. Do you know why this is?

**Reply by sgraband** [reply]
I investigated a bit and it seems like this is the expected behavior for kubernetes. K8s will not enforce this limit, it is basically just a label and its up to the cloud provider to enforce this setting. I would assume, that you are working on minikube, where this is probably not implemented. Hope this answers your question.

**Reply by wyhdil** [reply]
Okay, thank you for your response!
